docs/writing-policies/go/04-validation.md

```diff
@@ -8,331 +8,420 @@ title: ""
+FAIL
+exit status 1
+FAIL	github.com/kubewarden/go-policy-template	0.002s
+make: *** [Makefile:24: test] Error 1
+```
 
-	if response.Accepted != false {
-		t.Error("Unexpected accept response")
-	}
+As we can see all the `Settings` tests are passing, but there's one test case of the
+`TestValidateLabel` that is not:
 
-	expected_message := "Label owner is on the deny list"
-	if response.Message != expected_message {
-		t.Errorf("Got '%s' instead of '%s'", response.Message, expected_message)
-	}
-}
+```console
+validate_test.go:115: Unexpected acceptance with test case
+{map[owner:team-kubewarden] [hello] map[cc-center:team-\d+] false}
 ```
 
-The next test ensures a request is rejected when one of the user defined
-constraints is not satisfied:
+In this scenario our policy settings dictate that Pods must have a label with
+key `cc-center` that satisfies the `team-\d+` regular expression.
+The Pod being tested doesn't have this label, hence it should be rejected. This
+isn't happening however.
 
-```go
-func TestRejectionBecauseConstrainedLabelNotValid(t *testing.T) {
-	constrainedLabels := make(map[string]*RegularExpression)
-	re, err := CompileRegularExpression(`^cc-\d+$`)
-	if err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
-	constrainedLabels["cc-center"] = re
+> **Note:** you might be wondering why the output of the unit tests features
+> lines like `NATIVE: |{"level":"debug","message":"validating pod object","name":"test-pod","namespace":"default"}`
+>
+> This output is produced by the `logger` statements used inside of the policy. This happens
+> only when the code is run outside of the WebAssembly context. This does not happen when
+> the policy is evaluated by Kubewarden, in the context the `logger` statements will emit
+> OpenTelemetry events instead.
 
-	settings := Settings{
-		DeniedLabels:      mapset.NewThreadUnsafeSetFromSlice([]interface{}{}),
-		ConstrainedLabels: constrainedLabels,
-	}
+### Fix the broken unit test
 
-	payload, err := kubewarden_testing.BuildValidationRequest(
-		"test_data/ingress.json",
-		&settings)
-	if err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
+To fix the broken test we just discovered we have to do a small change inside
+of our validation function.
 
-	responsePayload, err := validate(payload)
-	if err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
+Currently the core of our validation logic is made by the following lines:
 
-	var response kubewarden_testing.ValidationResponse
-	if err := json.Unmarshal(responsePayload, &response); err != nil {
-		t.Errorf("Unexpected error: %+v", err)
+```go
+for label, value := range pod.Metadata.Labels {
+	if err := validateLabel(label, value, &settings); err != nil {
+		return kubewarden.RejectRequest(
+			kubewarden.Message(err.Error()),
+			kubewarden.NoCode)
 	}
+}
+```
 
-	if response.Accepted != false {
-		t.Error("Unexpected accept response")
-	}
+Here we iterate over each label and make sure this isn't neither denied, nor
```

Comment:
Are we saying it isn't denied and it not violating one of the constraints specified by the user?
```suggestion
Here we iterate over each label and make sure this isn't neither denied, nor
```

Reply:
yes, exactly

Reply:
In which case, I'd suggest the below:

```suggestion
Here we iterate over each label and make sure this is neither denied, nor
```